### ISSUE TITLE
Clean getPossibleMutations

### ISSUE THREAD

**Issue by christianechevarria** [liquidcarrot/carrot#67]
![image](https://user-images.githubusercontent.com/23618650/57595477-3ffec380-7513-11e9-97b3-eaf47150f09a.png)

**Comment by christianechevarria** [comment]
After some testing and reworking to speed things up, it seems that using network.evolve without `efficientMutation` is consistently faster than with it.

`efficientMutation` relies on `getPossibleMutations` to weed out impossible mutations before they happen so that mutations are not attempted only to fail, but due to built-in checks already present in `evolve` and `mutate` the additional overhead of running another set of tests is not worth it.

There can still be an optimization here though. There are mutations that are attempted and then fail, this functionally reduces the mutation rate. We should look into building a solution for `mutate` which will continue to attempt mutations until a suitable one is discovered.

**Comment by christianechevarria** [comment]
Removed `getPossibleMutations` in 2d755077d76d38058601ada544aa10d073df5a64

Will circle back to rework `Network.selectMutationMethod` and `Network.mutate`.

Currently thinking about:

Reworking `selectMutationMethod` into another method that takes a potential mutation as a parameter and returns truthy output if possible and falsy otherwise.

And `mutate` would also be reworked into a function that takes the allowed mutation methods as a parameter, calls `selectMutationMethod` internally until a suitable method is found, and then handles the details of applying the mutation.

Still playing around with ideas for this, would love to hear your thoughts 👍

**Comment by dan-ryan** [comment]
Some other ideas in the original pull request: https://github.com/wagenaartje/neataptic/pull/79

**Comment by christianechevarria** [comment]
Just implemented something in ebfbc781a92e0df37b728058bb5b67e7dfcff825 that's pretty similar to one of the suggestions in the original pull request: https://github.com/wagenaartje/neataptic/pull/79

Here's some of the relevant code:
```
  selectMutationMethod: function (genome, allowedMutations, efficientMutation) {
    
    if(efficientMutation) {
      let filtered = allowedMutations ? [...allowedMutations] : [...this.mutation]
      let success = false
      while(!success) {
        const currentMethod = filtered[Math.floor(Math.random() * filtered.length)]
        
        if(currentMethod === methods.mutation.ADD_NODE && genome.nodes.length >= this.maxNodes || currentMethod === methods.mutation.ADD_CONN && genome.connections.length >= this.maxConns || currentMethod === methods.mutation.ADD_GATE && genome.gates.length >= this.maxGates) {
          success = false
        } else {
          success = genome.mutate(currentMethod)
        }
        
        // we're done
        if(success || !filtered || filtered.length === 0) return
        
        // if not, remove the impossible method
        filtered = filtered.filter(function(value, index, array) {
          return value.name !== currentMethod.name
        })
      }
    }
...
  }
```

Results:
Using `efficientMutation` is not always faster than a run without it (because of the stochastic nature of inital genome fitness) but it **definitely** produces much more consistent results inside of `Network.evolve()` 

efficientMutation also _disproportionately_ pays off as the mutation rate is set higher. Here is a screenshot of training for XOR with efficientMutation vs without -- when the mutation rate is set to `0.9`:

![Screenshot 2019-05-19 at 1 00 00 PM](https://user-images.githubusercontent.com/23618650/57985704-e548f880-7a39-11e9-84b1-fb2f3e8f6994.png)

Despite this, more rigorous benchmarks would be needed to figure out exactly how much faster `efficientMutation` works than the classic way

**Comment by dan-ryan** [comment]
> efficientMutation also disproportionately pays off as the mutation rate is set higher
Interesting.

I've just added some comments. The `selectMutationMethod` doesn't make much sense in the state it is currently in.
